### DISCUSSION TITLE
Consider adding support for flat config as JSON file

### DISCUSSION THREAD

**Discussion by mrazauskas** [eslint/eslint#18129] [General] [answered]
Are there any plans allowing `eslint.config.json` as the flat config? (Sorry, if this is a duplicate. Can’t find similar discussion around.)

Many arguments for using JSON config files were mention in #17853

To add one more: `.js` files are very often type checked. I was trying to use flat config in rather small repo. First I had to install `@eslint/js`, `@types/eslint__js` (two more packages!); next it appeared two of three plugins support flat config, but do not ship typings (possible those will ship more of `@types` packages soon). Third one (`eslint-typescript`) currently does not support `moduleResolution: 'node'`, so it has only partial TypeScript support.

Type checking does not work at this moment. `// @ts-ignore` is the best way forward ;D

Note that relying on `@types` package means that types will be always behind. Users will have to match package versions, that means more documentation, more complexity. Felt like flat config was about simplicity.

Allowing JSON config file solves all these problems. I understand the novelty of `eslint.config.js`. It is attractive and fun. After the dust will settle, could you consider supporting old good JSON, please? It is not that I am asking to bring back legacy config system. All I want is support for flat config as JSON file. In some elegant way.

**Comment by mdjermanovic** [comment]
> Are there any plans allowing `eslint.config.json` as the flat config?

No, because it wouldn't be possible to use plugins or custom parsers in a JSON config file.

**Reply by mrazauskas** [reply]
It sounds not fair to say "wouldn't be possible". There are many JavaScript tolls which support JSON configs and are able to pass options to those plugins.

I do understand there is no interested to ~~implement~~ think about this right now. Honestly, I am not asking for that. There is no urgency.

**Reply by mdjermanovic** [reply]
I meant not possible because in the new config system ESLint by design doesn't load plugins and parsers, so you can't specify a plugin/parser by name but you have to provide plugin/parser objects, which cannot be done in a JSON file. One of the main design goals was to avoid module resolution, as this was the cause of many problems for users in the eslintrc configuration system.

(please see https://github.com/eslint/rfcs/tree/main/designs/2019-config-simplification for more details)

**Reply by mrazauskas** [reply]
It is all nicely thought through and interesting to read. The main problem I am pointing to: type checking does not work.

Maintainers of plugins are not used to ship types. Is there any guidance? Recommendations? I could go around and open issues in each repo pointing to best practises. Only that I can’t find any guide or discussion. Are any resources? Some centralised resource would do better job than wrestling with each and every maintainer separately.

The document you are pointing to is titled "Config File Simplification". This is exactly what I try to say in the OP. Type checking makes the new config system complex (additional packages, missing types, partial resolution support, etc). This was not an issue with JSON config. Might be there is other solutions. `eslint.config.json` was just a quick idea.

**Comment by mrazauskas** [comment, marked as answer]
I got an email saying "Closed as resolved". To be honest, this should have label: "Closed as ignored" ;D
